controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +139,317 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+	if existing.UID != "" {
+
+		// Since we have to update the existing StorageClass, so we will delete the existing storageclass and create a new one.
+		r.log.Info("StorageClass needs to be updated, deleting it.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+
+		// Check if StorageClass is in use, if yes, then fail.
+		// Wait until all PVs using the StorageClass under deletion are removed.
+		// Check for any PVs using the StorageClass.
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+		}
+
+		// Remove finalizer from StorageClass.
+		storageClass.SetFinalizers(remove(storageClass.Finalizers, StorageClassFinalizer))
+		if err := r.update(storageClass); err != nil {
+			return fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+		}
+
+		// Delete the StorageClass.
+		err = r.delete(existing)
+		if err != nil {
+			r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+			return err
+		}
+	}
+	r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+	err := r.Client.Create(r.ctx, storageClass)
+	if err != nil {
+		return fmt.Errorf("failed to create StorageClass: %v", err)
+	}
+	return nil
+}
+
+func (r *StorageClassClaimReconciler) reconcileProviderPhases() (reconcile.Result, error) {
+	log := ctrllog.FromContext(r.ctx)
 	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
 	return reconcile.Result{}, nil
 }
 
 func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&ocsv1alpha1.StorageClassClaim{}).
+		For(&ocsv1alpha1.StorageClassClaim{}, builder.WithPredicates(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// Validation check.
+				oldStorageClassClaim := e.ObjectOld.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaim := e.ObjectNew.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaimKRef := klog.KRef(newStorageClassClaim.Namespace, newStorageClassClaim.Name)
+				switch {
+				case oldStorageClassClaim.Spec.Type != newStorageClassClaim.Spec.Type:
+					r.log.Info("StorageClassClaim type change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				case oldStorageClassClaim.Spec.EncryptionMethod != newStorageClassClaim.Spec.EncryptionMethod:
+					r.log.Info(
+						"StorageClassClaim encryption method change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				default:
+					return true
+				}
+			},
```

Comment:
Why are we trying to validate for claim changes here? 
We need to validate against the actual storage class that is on the cluster (the actual state) which can only happen in the reconciler function.

Reply:
When claim changes, StorageClass changes. Is there a claim field that doesn't affect StorageClass?

Validating StorageClass is just delaying validation until we hit reconcile. This avoids calling reconcile as we know it will definitely fail.

Since both approaches will end up logging error and moving on, I don't have any preferences here.

Reply:
ACK

Reply:
@umangachapagain The issue I have with using predicates for this kind of check is the fact that the predicate cannot affect the status of the reconciled object. 

What I would like to see is a `validating` phase (reflected at the `Status.Phase` field at the start of the reconciler, and a `failed` state in case the desired state cannot be fulfilled as of reasons of wrong user input (like changing the type of an existing claim) - of course not re-queuing a reconcile in this case. 

Filtering via a predicate will not achieve any of the above